File: src/std/readFile.c
#ifndef READ_FILE_H
#define READ_FILE_H

#include "Allocator.c"
#include "Result.h"
#include "defs.h"
#include "readAllAlloc.c"
#include <stdio.h>

#ifdef BUILDING_WITH_ZIG
extern Result(Slice_char) readFile(Allocator ally, const char *path);
#else
static Result(Slice_char) readFile(Allocator ally, const char *path) {
  FILE *f = fopen(path, "r");
  if (!f) {
    return Result_Err(Slice_char, "could not open file");
  }
  Result(Slice_char) res = readAllAlloc(ally, f);
  if (fclose(f)) {
    if (res.ok)
      resizeAllocation(ally, char, &res.val, 0);
    return Result_Err(Slice_char, "could not close file");
  }
  return res;
}
#endif

#endif /* READ_FILE_H */

File: src/std/readFile.zig
const std = @import("std");

const Slice = @import("Slice.zig").Slice;
const Result = @import("Result.zig").Result;
const alloc = @import("Allocator.zig");
const Allocator = alloc.Allocator;
const readAllAlloc = @import("readAllAlloc.zig").readAllAlloc;

const Res = Result(Slice(u8));

pub export fn readFile(ally: Allocator, path: [*:0]const u8) Res {
    const f = std.c.fopen(path, "r") orelse return Res.Err("could not open file");
    var res = readAllAlloc(ally, f);
    if (std.c.fclose(f) != 0) {
        if (res.ok) alloc.resizeAllocation(ally, u8, &res.x.val, 0);
        return Res.Err("could not close file");
    }
    return res;
}
